### PR TITLE
Add failures to partition status cache

### PR DESCRIPTION
**Pull request by johannkm**

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12333** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12333" target="_blank">Graphite</a> 
    * **PR #12404** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12404" target="_blank">Graphite</a> 
      * **PR #12405** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12405" target="_blank">Graphite</a> 
      * **PR #12599** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12599" target="_blank">Graphite</a>  👈
      * **PR #12570** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12570" target="_blank">Graphite</a> 
    * **PR #12310** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12310" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12599?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/storage/partition_status_cache.py [hunk 1]
```diff
@@ -197,6 +219,80 @@ def _build_status_cache(
...
+        get_validated_partition_keys(instance, partitions_def, failed_partitions)
+    )
+
+
+def _get_updated_failed_partition_subset(
+    instance: DagsterInstance,
+    asset_key: AssetKey,
+    partitions_def: PartitionsDefinition,
+    current_cached_subset: PartitionsSubset,
+    unevaluated_planned_event_records: Sequence[EventLogRecord],
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/storage/partition_status_cache.py:268]: can we rename this arg to unevaluated_event_records? since it contains both materialization and materialization planned events

File: python_modules/dagster/dagster/_core/storage/partition_status_cache.py [hunk 2]
```diff
@@ -239,24 +345,42 @@ def _get_updated_status_cache(
...
-                unevaluated_materializations.event_log_entry.dagster_event.partition
-            )
+    for record in unevaluated_event_records:
+        if not record.event_log_entry.dagster_event:
+            check.failed("Expected dagster event")
+
+        if record.event_log_entry.dagster_event.is_step_materialization:
+            if record.event_log_entry.dagster_event.partition:
+                newly_materialized_partitions.add(record.event_log_entry.dagster_event.partition)
+        elif record.event_log_entry.dagster_event.is_asset_materialization_planned:
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/storage/partition_status_cache.py:355]: think this might need to be removed?

File: python_modules/dagster/dagster_tests/storage_tests/test_asset_status_cache.py [hunk 3]
```diff
@@ -316,3 +316,44 @@ def asset1():
...
+    @asset(partitions_def=partitions_def)
+    def asset1(context):
+        if context.partition_key.startswith("fail"):
+            raise Exception()
+
+    asset_key = AssetKey("asset1")
+    asset_graph = AssetGraph.from_assets([asset1])
+    asset_job = define_asset_job("asset_job").resolve([asset1], [])
+
+    with instance_for_test() as created_instance:
```
Comment (clairelin135) [inline python_modules/dagster/dagster_tests/storage_tests/test_asset_status_cache.py:333]: can we also test what the cache value looks like when no events exist?

I think it's also worth adding a test for the unpartitioned case, to assert that both the failed subset and materialized subset are `None`. Those values should only be `empty_subset`s when unpartitioned, or partitioned but never updated.

File: python_modules/dagster/dagster_tests/storage_tests/test_asset_status_cache.py [hunk 4]
```diff
@@ -316,3 +316,44 @@ def asset1():
         )
         assert cached_status
         assert cached_status.serialized_materialized_partition_subset is None
+
+
+def test_failure_cache():
```
Comment (clairelin135) [inline python_modules/dagster/dagster_tests/storage_tests/test_asset_status_cache.py:321]: can we also add another test that contains a previously serialized asset status cache value that doesn't contain the failed partitions subset? and then confirm that we can deserialize that object and that upon updating we'll also have a failed partitions subset in it?

I think this behavior currently works in your PR, but I think it would be helpful for confirmation since this is the code path that all our current users will hit.

File: python_modules/dagster/dagster/_core/storage/partition_status_cache.py [hunk 5]
```diff
@@ -94,6 +108,14 @@ def deserialize_materialized_partition_subsets(
 
         return partitions_def.deserialize_subset(self.serialized_materialized_partition_subset)
 
+    def deserialize_failed_partition_subsets(
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/storage/partition_status_cache.py:110]: it's used in https://github.com/dagster-io/dagster/pull/12570
